ntoskrnl/fsrtl/fastio.c

```diff
@@ -1613,6 +1630,46 @@ FsRtlAcquireFileExclusiveCommon(IN PFILE_OBJECT FileObject,
             return Status;
         }
     }
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+NTAPI
+FsRtlAcquireFileExclusiveCommon(IN PFILE_OBJECT FileObject,
+                                IN FS_FILTER_SECTION_SYNC_TYPE SyncType,
+                                IN ULONG Reserved)
+{
+    PFSRTL_COMMON_FCB_HEADER FcbHeader;
+    PDEVICE_OBJECT DeviceObject;
+    PFAST_IO_DISPATCH FastDispatch;
+    PEXTENDED_DRIVER_EXTENSION DriverExtension;
+    PFS_FILTER_CALLBACKS FilterCallbacks;
+    FS_FILTER_PARAMETERS FilterParameters;
+
+    /* Get Device Object and Fast Calls */
+    FcbHeader = (PFSRTL_COMMON_FCB_HEADER)FileObject->FsContext;
+    DeviceObject = IoGetRelatedDeviceObject(FileObject);
+
+    /* Get master FsRtl lock */
+    FsRtlEnterFileSystem();
+
+    DriverExtension = (PEXTENDED_DRIVER_EXTENSION)DeviceObject->DriverObject->DriverExtension;
+    FilterCallbacks = DriverExtension->FsFilterCallbacks;
+    if (FilterCallbacks)
+    {
+        RtlZeroMemory(&FilterParameters, sizeof(FilterParameters));
+        FilterParameters.AcquireForSectionSynchronization.PageProtection = Reserved;
+        FilterParameters.AcquireForSectionSynchronization.SyncType = SyncType;
+
+        /* Call pre/post callbacks */
+        return FsRtlpCallPrePostCallbacks(DeviceObject,
+                                          FileObject,
+                                          FS_FILTER_ACQUIRE_FOR_SECTION_SYNCHRONIZATION,
+                                          FilterCallbacks,
+                                          &FilterParameters,
+                                          FilterCallbacks->PreAcquireForSectionSynchronization,
+                                          FilterCallbacks->PostAcquireForSectionSynchronization);
```

Comment:
The `Pre` callback must be invoked **before** the operation, the `Post` callback **after** it has completed. When the `Pre` callback returns an error, the function should fail.

See https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-fs_filter_callbacks

> The filter notification callback routine is invoked before the operation request is passed to lower-level filter drivers and the underlying file system. In the callback routine, the filter driver should perform any needed processing and immediately return STATUS_SUCCESS. If a filter driver's callback routine returns a status value other than STATUS_SUCCESS, this causes the operation request to fail.

> The filter completion callback routine is invoked after the operation request is passed to lower-level filter drivers and the underlying file system.

Reply:
OK, just updated the code by new changes. The new pushed implementation is more similar to what Windows does. The post callback should be executed after other things are done, at the end of the function. BUT. FsFilter callbacks and other resource acquire/release calls should be executed separately, in the different cases: the normal calls - for base FS drivers, and pre and post filter callbacks - for FsFilter drivers. This is the nuance which isn't explained in the MSDN document you're linked above, but it's confirmed by my previous analysis as well, and the outcome results of the tests are proving this too: it works properly with the current code, but does not if after executing a Pre callback, simply fall to the normal acquire/release calls, then to Post and return. In such a case, it asserts, excepts and has another failures when browsing some FS partitions like CDFS and network drives. So note that I added an appropriate gotos in the code, to make it working correctly. This was already discussed by me and Hermes in PM if what.

Reply:
The usage of either base FS driver or Filter driver can be tested by checking whether `DriverExtension->FsFilterCallbacks` member is valid or not (equals to `NULL`). It should be `NULL` in the 1st case, so then Pre and Post callbacks should not be executed at all but the normal acquire/release calls instead, and it needs to be valid in the 2nd case, when the **only** Filter callbacks should be executed, but other things shouldn't be called.

Reply:
You still need to call the base FS driver somehow. Just returning success as soon as the filter driver's callback was invoked, will break things.